DESIGN.md

```diff
@@ -0,0 +1,79 @@
+# RPC
+This project implements an RPC server in Go using gRPC as a transport.
+
+# Design
+What follows is a design document outlining the technology choices, architecture, and implementation of the RPC client and server. 
+
+## Authentication and Authorization
+The RPC service includes mutual TLS authentication. The client and server will both have their own private keys as well as certificates signed by the same CA. The minimum TLS version will be set to TLS1.2. In order to protect against downgrade attacks in TLS1.2, an allowlist will be used for ciphers. Conveniently, the Go TLS package will return a list of all cipher suites it implements excluding suites with known security issues (tls.CipherSuites()). In order to restrict the client to using these only, PreferServerCipherSuites will be set to true in the server TLS configuration. If this were a production application, an endpoint would be included in the API to accept a certificate revocation list, allowing the administrator to 'ban' certain clients. 
+
+For Authorization, a simple form of role based access control will be implemented: Each role will have its own CA. All CAs will be trusted by the server. Upon handling a remote procedure call, the server will first check the client certificate's certificate authority and compare it to its internal mapping of certificate authority to privilege level. The following roles will be implemented:
+- **Read only:** this role can get job output, but not create or delete jobs. This is useful for automated tooling such as status indicators and job notifiers.
+- **Read/Write access:** this role can get job output as well as create and delete jobs. This role is useful for administrators and any kind of deployment automation that would leverage this service.
+
+![visual representation of the multiple client CA model](doc/client_cas.png)
+
+## API
+### Endpoints
+The API exposed by the RPC service will contain the following endpoints
+- **CreateJob**: receives a Job and returns a CreateJobOutput
+- **DeleteJob**: receives a PID and returns a DeleteJobOutput
+- **ListJobs**: returns a JobList
+- **GetOutput**: receives a PID and returns a JobOutput
+
+### Datatypes
+- **Job**
+    - Command (string):
+    Command holds a valid shell command that is run on the underlying server shell.
+    - PID (integer):
```

Comment:
I am sorry I am a bit confused by this. 
when you say in the 'Endpoints' section above that the `CreateJob` enpoint will receive a `Job`, do you mean this struct? If yes how will you get the pid if the job hasn't been created yet? If you're not filling the struct on the client side, why not only send the command and let the server figure out the rest

Reply:
In the updated revision of this section the client transmits only the command to the server in the CreateJob call.